src/entities/trades/ecoRouterTrades/tradeTypes.ts

```diff
@@ -0,0 +1,313 @@
+import { AddressZero } from '@ethersproject/constants'
+import { Provider } from '@ethersproject/providers'
+
+import { ChainId, TradeType } from '../../../constants'
+import { Token } from '../../token'
+import { ZeroXTrade } from '../0x'
+import { CurveTrade } from '../curve'
+import { CoWTrade } from '../gnosis-protocol'
+import { Trade } from '../interfaces/trade'
+import { RoutablePlatform } from '../routable-platform'
+import { UniswapTrade } from '../uniswap'
+import { getAllCommonUniswapV2Pairs, UniswapV2Trade } from '../uniswap-v2'
+// Types
+import {
+  EcoRouterBestExactInParams,
+  EcoRouterBestExactOutParams,
+  EcoRouterResults,
+  EcoRouterSourceOptionsParams,
+} from './types'
+import { getUniswapV2PlatformList, sortTradesByExecutionPrice } from './utils'
+
+/**
+ * Low-level function to fetch from Eco Router sources
+ * @returns {Promise<EcoRouterResults>} List of unsorted trade sources
+ */
+export async function getExactIn(
+  { currencyAmountIn, currencyOut, maximumSlippage, receiver = AddressZero, user }: EcoRouterBestExactInParams,
+  { uniswapV2 }: EcoRouterSourceOptionsParams,
+  provider?: Provider
+): Promise<EcoRouterResults> {
+  // Error list
+  const errors: any[] = []
+  // Derive the chainId from the token in or out
+  const chainId = (currencyAmountIn.currency as Token).chainId ?? (currencyOut as Token).chainId
+
+  if (!chainId) {
+    return {
+      errors: [new Error('Unsupported chain')],
+      trades: [],
+    }
+  }
+
+  // Uniswap V2
+  // Get the list of Uniswap V2 platform that support current chain
+  const uniswapV2PlatformList = getUniswapV2PlatformList(chainId)
+
+  const uniswapV2TradesList = uniswapV2PlatformList.map(async (platform) => {
+    try {
+      const getAllCommonUniswapV2PairsParams = {
+        currencyA: currencyAmountIn.currency,
+        currencyB: currencyOut,
+        platform,
+        provider,
+      }
+
+      const pairs = await getAllCommonUniswapV2Pairs(getAllCommonUniswapV2PairsParams)
+
+      return (
+        UniswapV2Trade.computeTradesExactIn({
+          currencyAmountIn,
+          currencyOut,
+          maximumSlippage,
+          maxHops: {
+            maxHops: uniswapV2.useMultihops ? 3 : 1,
+            maxNumResults: 1,
+          },
+          pairs,
+        })[0] ?? undefined
+      )
+    } catch (error) {
+      errors.push(error)
+      return undefined
+    }
+  })
+
+  const uniswapTrade = new Promise<UniswapTrade | undefined>((resolve) => {
+    if (!RoutablePlatform.UNISWAP.supportsChain(chainId)) {
+      return resolve(undefined)
+    }
+
+    UniswapTrade.getQuote({
+      quoteCurrency: currencyOut,
+      amount: currencyAmountIn,
+      maximumSlippage,
+      recipient: receiver,
+      tradeType: TradeType.EXACT_INPUT,
+    })
+      .then((res) => resolve(res ? res : undefined))
+      .catch((error) => {
+        console.error(error)
+        errors.push(error)
+        resolve(undefined)
+      })
```

Comment:
We are repeating this logic every `.catch` maybe we can have a function catch error that receives and error and returns undefined (or just `return`). Calling `resolve` is the same as calling `return`, so we can simply that too.

Reply:
Hmm in not sure about this actually because its a promise not a function...